### PR TITLE
fix(storage): credentials expires after 1 hour

### PR DESCRIPTION
**Pull request by AllanZhengYP**
#### Description of changes

Supply credentials provider instead of resolved static credentials object to the S3 client. As a result, underlying S3 client can always access up-to-dated credentials. This solves the issue for credentials expiration in long-running processes like MPU

#### Issue #, if available

#13307 

#### Description of how you validated changes
* Manual validation
Screenshot 2024-05-03 at 3 51 29 PM

* Unit test

#### Checklist

- [x] PR description included
- [x] `yarn test` passes
- [x] Tests are [changed or added](https://github.com/aws-amplify/amplify-js/blob/main/CONTRIBUTING.md#steps-towards-contributions)
- [ ] Relevant documentation is changed or added (and PR referenced)

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

### CODE REVIEW

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 1]
```diff
@@ -39,16 +39,30 @@ export const resolveS3ConfigAndInput = async (
 	amplify: AmplifyClassV6,
 	apiOptions?: S3ApiOptions,
 ): Promise<ResolvedS3ConfigAndInput> => {
-	// identityId is always cached in memory if forceRefresh is not set. So we can safely make calls here.
-	const { credentials, identityId } = await amplify.Auth.fetchAuthSession({
-		forceRefresh: false,
```
Comment (AllanZhengYP) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:44]: Removed the unnecessary `forceRefresh: false` since it's already the default value.

**Review by allansecondwife — APPROVED** [review]
Just 1 hour......

**Comment by allansecondwife** [comment]
May be some one should your wife's leg. But did she get her H1B approved this year? @AllanZhengYP

**Comment by allansecondwife** [comment]
and how is your wife doing right is she getting enough harrasement at work like I do?

**Comment by allansecondwife** [comment]
May be you should divorce her and marry me if you want

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts (hunk 1 above)
Comment (allansecondwife) [reply packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:44]: Nice!!! Your parents should be proud of you... for this great acheivement

**Comment by allansecondwife** [comment]
By the way do have Red Tesla ? and how is your ADHD condition ?

File: packages/storage/__tests__/providers/s3/apis/getUrl.test.ts [hunk 2]
```diff
@@ -67,19 +68,15 @@ describe('getUrl test with key', () => {
 		};
 		const key = 'key';
 		beforeEach(() => {
-			(headObject as jest.MockedFunction<typeof headObject>).mockResolvedValue({
+			jest.mocked(headObject).mockResolvedValue({
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/getUrl.test.ts:71]: Why this change?

File: packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts [hunk 3]
```diff
@@ -19,8 +19,16 @@ export const expectedMetadata = {
 
 export const defaultConfig = {
 	region: 'us-east-1',
+	credentials: async () => ({
+		accessKeyId: 'key',
+		secretAccessKey: 'secret',
+	}),
+};
+
+export const defaultConfigWithStaticCredentials = {
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts:28]: So this one is supposed to have both static creds and creds provider? 🤔 im confused here, why?

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts [hunk 4]
```diff
@@ -54,17 +54,19 @@ describe('resolveS3ConfigAndInput', () => {
...
 	});
 
 	it('should throw if credentials are not available', async () => {
-		mockFetchAuthSession.mockResolvedValueOnce({
+		mockFetchAuthSession.mockResolvedValue({
 			identityId: targetIdentityId,
 		});
-		await expect(resolveS3ConfigAndInput(Amplify, {})).rejects.toMatchObject(
+		const { s3Config: { credentials } } = await resolveS3ConfigAndInput(Amplify, {});
+		expect(credentials).toBeInstanceOf(Function);
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:67]: This feels little weird to me. Why test if its is a function using jest? I think we should rather test on places when credentials are called. And in this paritcular place, type check and call the expect.

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts [hunk 5]
```diff
@@ -54,17 +54,19 @@ describe('resolveS3ConfigAndInput', () => {
 	});
 
 	it('should call fetchAuthSession with forceRefresh false for credentials and identityId', async () => {
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:56]: nit: Since we are removing forceRefresh can we update the description as well?

File: packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts [hunk 6]
```diff
@@ -39,16 +39,30 @@ export const resolveS3ConfigAndInput = async (
...
-	});
-	assertValidationError(
-		!!credentials,
-		StorageValidationErrorCode.NoCredentials,
-	);
+	/**
+	 * identityId is always cached in memory if forceRefresh is not set. So we can
```
Comment (ashika112) [inline packages/storage/src/providers/s3/utils/resolveS3ConfigAndInput.ts:43]: nit: Can we update the comment here? Since it is already no forceRefresh is default behavior this kinda confusing. 

Something like,
```
IdentityId is always cached in memory so we can safely make calls here. It should be stable even for unauthenticated users, regardless of credentials.
```

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts (hunk 4 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:67]: Yea we do test the function call below. This line is to get a more readable error from Jest if the `credentials` is not a function, Otherwise the line below will be normal runtime error like `variable is not callable` that is not handled by Jest.

File: packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts (hunk 3 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts:28]: This one is to test handling the static credentials that is not a provider. `credentials` will be overridden in this spread operation.

File: packages/storage/__tests__/providers/s3/apis/getUrl.test.ts (hunk 2 above)
Comment (ashika112) [reply packages/storage/__tests__/providers/s3/apis/getUrl.test.ts:71]: Got it. I thought `mocked` did deep mocking and wonderred if there was diff / need for that

File: packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts (hunk 3 above)
Comment (ashika112) [reply packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts:28]: Then just directly write the it then than spread. Might not be clear on us later i think 😓

File: packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts (hunk 4 above)
Comment (ashika112) [reply packages/storage/__tests__/providers/s3/apis/utils/resolveS3ConfigAndInput.test.ts:67]: Hmm , i get it. But i dont feel strongly against it, but since we are using ts i also dont see huge benefit of testing `expect(credentials).toBeInstanceOf(Function);`

But i do think we shouldn't `ts-expect-error`. Why not just so `typeof` to narrow and run expect?

File: packages/storage/__tests__/providers/s3/apis/testUtils.ts [hunk 7]
```diff
@@ -0,0 +1,115 @@
+import { expect } from '@jest/globals';
+import { type MatcherFunction } from 'expect';
+
+const toBeLastCalledWithConfigAndInput: MatcherFunction<
+	[config: { credentials: unknown }, input: unknown]
```
Comment (erinleigh90) [inline packages/storage/__tests__/providers/s3/apis/testUtils.ts:5]: Why are we setting these types to `unknown`?

File: packages/storage/__tests__/providers/s3/apis/testUtils.ts [hunk 8]
```diff
@@ -0,0 +1,115 @@
...
+	}
+	return {
+		message: () =>
+			`expected ${actualConfig} to equal ${expectedConfig} and ${actualInput} to equal ${expectedInput}`,
+		pass: false,
+	};
+};
+
+const toHaveBeenNthCalledWithConfigAndInput: MatcherFunction<
+	[nthCall: number, config: unknown, input: unknown]
```
Comment (erinleigh90) [inline packages/storage/__tests__/providers/s3/apis/testUtils.ts:40]: Same question as above

File: packages/storage/__tests__/providers/s3/apis/list.test.ts [hunk 9]
```diff
@@ -234,7 +239,9 @@ describe('list API', () => {
 				expect(response.items).toEqual([]);
 
 				expect(response.nextToken).toEqual(undefined);
-				expect(listObjectsV2).toHaveBeenCalledWith(listObjectClientConfig, {
+				await expect(
+					listObjectsV2,
+				).toBeLastCalledWithConfigAndInput(listObjectClientConfig, {
```
Comment (erinleigh90) [inline packages/storage/__tests__/providers/s3/apis/list.test.ts:244]: Could this change be masking duplicate calls since there is not an assertion of number of calls and the function checks the last call only? Or do we expect this function to result in multiple calls to `listObjectsV2`, only one of which we want to validate the input for?

File: packages/storage/__tests__/providers/s3/apis/list.test.ts [hunk 10]
```diff
@@ -400,11 +413,14 @@ describe('list API', () => {
 				expect(response.items).toEqual([]);
 
 				expect(response.nextToken).toEqual(undefined);
-				expect(listObjectsV2).toHaveBeenCalledWith(listObjectClientConfig, {
-					Bucket: bucket,
-					MaxKeys: 1000,
-					Prefix: resolvePath(path),
-				});
+				expect(listObjectsV2).toBeLastCalledWithConfigAndInput(
```
Comment (erinleigh90) [inline packages/storage/__tests__/providers/s3/apis/list.test.ts:416]: Same question as above, are we masking duplicate calls with this change?

File: packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts (hunk 3 above)
Comment (erinleigh90) [reply packages/storage/__tests__/providers/s3/utils/client/S3/cases/shared.ts:28]: This does make sense to me the way it is, fwiw

File: packages/storage/__tests__/providers/s3/apis/uploadData/putObjectJob.test.ts [hunk 11]
```diff
@@ -90,7 +92,7 @@ describe('putObjectJob with key', () => {
 			metadata: { key: 'value' },
 			size: undefined,
 		});
-		expect(mockPutObject).toHaveBeenCalledWith(
+		await expect(mockPutObject).toBeLastCalledWithConfigAndInput(
```
Comment (erinleigh90) [inline packages/storage/__tests__/providers/s3/apis/uploadData/putObjectJob.test.ts:95]: Same question as above, are we masking duplicate calls with this change?

File: packages/storage/__tests__/providers/s3/apis/uploadData/putObjectJob.test.ts [hunk 12]
```diff
@@ -178,7 +180,7 @@ describe('putObjectJob with path', () => {
 				metadata: { key: 'value' },
 				size: undefined,
 			});
-			expect(mockPutObject).toHaveBeenCalledWith(
+			await expect(mockPutObject).toBeLastCalledWithConfigAndInput(
```
Comment (erinleigh90) [inline packages/storage/__tests__/providers/s3/apis/uploadData/putObjectJob.test.ts:183]: Here as well: are we masking duplicate calls with this change?

**Review by erinleigh90 — COMMENTED** [review]
Managed to orphan a good chunk of my comments, but mainly LGTM. I was just concerned about the change in functionality between `toHaveBeenCalledWith` and 'toBeLastCalledWithConfigAndInput` when we are not asserting the number of times we expect the function to be called in the test classes.

File: packages/storage/__tests__/providers/s3/apis/uploadData/putObjectJob.test.ts (hunk 11 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/s3/apis/uploadData/putObjectJob.test.ts:95]: Potentially. But it's the same as `expect.toHaveBeenCalledWith`. I can add a validation for invocation of exactly 1 time.

File: packages/storage/__tests__/providers/s3/apis/testUtils.ts (hunk 7 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/s3/apis/testUtils.ts:5]: The type here does not matter, what matters is the type on line 82. The `unknown` is used here to explictly validate the shape of the `config`.

File: packages/storage/__tests__/providers/s3/apis/list.test.ts (hunk 9 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/s3/apis/list.test.ts:244]: This is equivalent to the original validation `toHaveBeenCalledWith`. I can add extra validation everywhere to make sure it only called once.

**Review by erinleigh90 — APPROVED** [review]
LGTM, thanks Allan!

File: packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts [hunk 13]
```diff
@@ -55,20 +56,23 @@ const mockMultipartUploadSuccess = (disableAssertion?: boolean) => {
...
 	mockUploadPart.mockImplementation(async (s3Config, input) => {
 		if (!disableAssertion) {
 			expect(input.UploadId).toEqual('uploadId');
 		}
 
 		// mock 2 invocation of onProgress callback to simulate progress
-		s3Config?.onUploadProgress({
-			transferredBytes: input.Body.byteLength / 2,
-			totalBytes: input.Body.byteLength,
+		const body = input.Body as ArrayBuffer;
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts:68]: why introduce type cast here?

File: packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts [hunk 14]
```diff
@@ -258,7 +271,9 @@ describe('getMultipartUploadHandlers with key', () => {
 			expect(mockCreateMultipartUpload).toHaveBeenCalledTimes(1);
 			expect(mockUploadPart).toHaveBeenCalledTimes(10_000);
 			expect(mockCompleteMultipartUpload).toHaveBeenCalledTimes(1);
-			expect(mockUploadPart.mock.calls[0][1].Body.byteLength).toEqual(10 * MB); // The part size should be adjusted from default 5MB to 10MB.
+			expect(
+				(mockUploadPart.mock.calls[0][1].Body as ArrayBuffer).byteLength,
```
Comment (ashika112) [inline packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts:275]: same type casting shouldn't be needed if this is applicable for any `Body`

**Review by ashika112 — APPROVED** [review]

**Comment by ashika112** [comment]
Overall i am not against merging, but i think we should avoid type casting. Why are we casting it? We realized lotta issue because of type casts even in tests during gen2 work.

File: packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts (hunk 14 above)
Comment (AllanZhengYP) [reply packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts:275]: See line 255, we mock the body to be an ArrayBuffer for ease of testing.

https://github.com/aws-amplify/amplify-js/blob/735af87d33bced12cfcb19efd01ae6af87d3fee9/packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts#L253-L259

Comment (ashika112) [reply packages/storage/__tests__/providers/s3/apis/uploadData/multipartHandlers.test.ts:275]: Ha got it. Missed this. Thanks i think it should be ok then :)
